### PR TITLE
Remove all binding results ranking and use all found definitions instead

### PR DESCRIPTION
**Pull request by ggiraldez**
This PR removes `resolve_definition()` and consequently all the ranking machinery associated with it, including C3 linearisation and bindings context, which was used for virtual method lookups. Also deprecates binding assertions in favour of straight snapshots. This simplifies the binding resolution code quite a bit.

### CODE REVIEW

File: crates/metaslang/bindings/src/lib.rs [hunk 1]
```diff
@@ -525,28 +429,14 @@ impl<'a, KT: KindTypes + 'static> Reference<'a, KT> {
...
-        Resolver::build_for(self, ResolveOptions::NonRecursive).first()
-    }
-
-    pub(crate) fn has_tag(&self, tag: Tag) -> bool {
-        self.owner
-            .references_info
-            .get(&self.handle)
-            .and_then(|info| info.tag)
-            .is_some_and(|reference_tag| reference_tag == tag)
+        Resolver::build_for(self, ResolveOptions::NonRecursive).all()
```
Comment (OmarTawfik) [inline crates/metaslang/bindings/src/lib.rs:439]: I wonder if the comment here needs to be updated? Not sure I'm following why the change to `all()`.

File: crates/solidity/testing/perf/src/tests/references.rs [hunk 2]
```diff
@@ -17,16 +17,16 @@ pub fn run(binding_graph: BindingGraph) {
...
+        if !definitions.is_empty() {
             resolved_references += 1;
         }
     }
 
     assert_eq!(reference_count, 1652, "Failed to fetch all references");
 
     assert_eq!(
-        resolved_references, 1409,
+        resolved_references, 1490,
```
Comment (OmarTawfik) [inline crates/solidity/testing/perf/src/tests/references.rs:29]: I wonder why would the # of references go up in this case?

File: crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt [hunk 3]
```diff
@@ -21,10 +21,10 @@ References and definitions:
     │           ╰──── ref: 4
   7 │         tx;
     │         ─┬  
-    │          ╰── ref: 2
+    │          ╰── refs: 2, built-in
   8 │         sha256();
```
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt:25]: Looks like all functions with the same name are returned now, rather than resolving the closest one. Is this expected?

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left a couple of questions.

File: crates/solidity/testing/perf/src/tests/references.rs (hunk 2 above)
Comment (ggiraldez) [reply crates/solidity/testing/perf/src/tests/references.rs:29]: Before we were checking if the reference resolved to a single definition. Now we're checking if it resolves to *any* number of definitions. This difference is the number of previously ambiguous resolution results.

File: crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt (hunk 3 above)
Comment (ggiraldez) [reply crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt:25]: This was previously resolved with the ranking algorithm applied to all results we got from the stack graph. For this case, built-ins were being de-prioritized by ranking. In all other cases, those are either virtual methods which were ranked using C3 linearisation, or import aliases.

File: crates/metaslang/bindings/src/lib.rs (hunk 1 above)
Comment (ggiraldez) [reply crates/metaslang/bindings/src/lib.rs:439]: Since we're no longer ranking the results, it makes no sense to pick the first result arbitrarily. This errs on the safe side and returns all results. This should not be a problem in the general case, since this non-recursive resolution should be applied to identifiers that are normally non-ambiguous (eg. contract names).

File: crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt:25]: Returning multiple method overloads for a method call is useful for many scenarios, and they are all correct. If an IDE does a rename, they all should be renamed. We consider them the same one.
But not sure this reasoning applies to this case. Here, the built-in is shadowed by the contract function. They are not the same definition. They are two separate ones that can be renamed/moved separately.

I assume this would have been handled by scopes, rather than ranking?

1. The global scope has access to the built-in `sha256` function.
2. The outer scope of the file adds the global `sha256` function.
3. The inner scope of the contract adds the private `sha256` function.
4. Name resolution of the `sha256()` method call inside `test()` should pick #3, and stop there. Right?

Why is it different than shadowed variables, or how locally-defined variables should shadow variables brought in from an `import * from "...."` statement?

Comment (ggiraldez) [reply crates/solidity/testing/snapshots/bindings_output/built_ins/shadowing/generated/0.7.1-success.txt:25]: Yeah, you're right. This case should be handled by scoping. We're missing setting  precedence in edges for in contract definitions. I'll fix this.

**Review by OmarTawfik — APPROVED** [review]
